dss-pades-pdfbox/src/main/java/eu/europa/esig/dss/pdf/pdfbox/PdfBoxDocumentReader.java

```diff
@@ -94,37 +99,69 @@ public class PdfBoxDocumentReader implements PdfDocumentReader {
+			throws IOException, InvalidPasswordException {
 		this(dssDocument, null);
 	}
 
 	/**
-	 * The PDFBox implementation of the Reader
-	 * 
-	 * @param dssDocument        {@link DSSDocument} to read
-	 * @param passwordProtection {@link String} a password to open a protected document
-	 * @throws IOException       if an exception occurs
-	 * @throws eu.europa.esig.dss.pades.exception.InvalidPasswordException if the password is not provided or
-	 *                           invalid for a protected document
+	 * Constructor of the PDFBox implementation of the Reader with a password protection phrase provided.
+	 * This constructor uses in-memory processing of the document.
+	 *
+	 * @param dssDocument           {@link DSSDocument} to read
+	 * @param passwordProtection    {@link String} a password to open a protected document
+	 * @throws IOException          if an exception occurs
+	 * @throws InvalidPasswordException if the password is not provided or
+	 *                              invalid for a protected document
 	 */
 	public PdfBoxDocumentReader(DSSDocument dssDocument, String passwordProtection)
-			throws IOException, eu.europa.esig.dss.pades.exception.InvalidPasswordException {
+			throws IOException, InvalidPasswordException {
+		this(dssDocument, passwordProtection, PAdESUtils.DEFAULT_PDF_MEMORY_USAGE_SETTING);
+	}
+
+	/**
+	 * Constructor of the PDFBox implementation of the Reader with a password protection phrase
+	 * and memory usage settings provided.
+	 * 
+	 * @param dssDocument           {@link DSSDocument} to read
+	 * @param passwordProtection    {@link String} a password to open a protected document
+	 * @param pdfMemoryUsageSetting {@link MemoryUsageSetting} PDF loading setting
+	 * @throws IOException          if an exception occurs
+	 * @throws InvalidPasswordException if the password is not provided or
+	 *                              invalid for a protected document
+	 */
+	public PdfBoxDocumentReader(DSSDocument dssDocument, String passwordProtection, PdfMemoryUsageSetting pdfMemoryUsageSetting)
+			throws IOException, InvalidPasswordException {
 		Objects.requireNonNull(dssDocument, "The document must be defined!");
+		Objects.requireNonNull(pdfMemoryUsageSetting, "PdfMemoryUsageSetting must be defined!");
 		this.dssDocument = dssDocument;
-		try (InputStream is = dssDocument.openStream()) {
-			this.randomAccessRead = new RandomAccessReadBuffer(is);
-			this.pdDocument = Loader.loadPDF(randomAccessRead, passwordProtection);
-		} catch (InvalidPasswordException e) {
-			throw new eu.europa.esig.dss.pades.exception.InvalidPasswordException(
-					String.format("Encrypted document : %s", e.getMessage()));
+		MemoryUsageSetting memoryUsageSetting = PdfBoxUtils.getMemoryUsageSetting(pdfMemoryUsageSetting);
+		try {
+			if (PdfMemoryUsageSetting.Mode.MEMORY_FULL != pdfMemoryUsageSetting.getMode() && dssDocument instanceof FileDocument) {
```

Comment:
The MEMORY_FULL could be used with a FileDocument (meaning that willingness is to fully load into memory the file) and I understand that you achieve that also with the stream and RandomAccessReadBuffer later in the code.
I just see it more complex and less linear, why not leave to the library? (that does the same anyway)

Reply:
Because it is not really the same. See the comparison between "Memory Full" and "Memory Buffered" in #177. There is a consistent difference in execution, mainly on validation and LTA-level extension.

Reply:
Sorry, I haven't caught that well.

I am surprised, the code is exactly the same for pdfbox:
`loadPDF(File file, String password,
            StreamCacheCreateFunction streamCacheCreateFunction)` 
--> `Loader.loadPDF(file, password, null, null, streamCacheCreateFunction)`
-->  https://github.com/apache/pdfbox/blob/357a2c14762f3e08f4ef48852d79f541ccaca6af/pdfbox/src/main/java/org/apache/pdfbox/Loader.java#L223

Reply:
While it uses the IOUtils.createMemoryOnlyStreamCache(), but creates a RandomAccessReadBufferedFile(file) instead of converting the document to binaries directly. I think it makes the difference. Possible that the file in filesystem is being accessed multiple number of times? Especially provided that there is a huge difference in openPdf implementation, I think it would be beneficial to keep the same logic in both implementations. Moreover, keeping the old behavior we reduce the possibility of regression, in case we miss something or PdfBox unexpectedly changes the handling inside.

Reply:
I see your point, I agree with you. It's safer to keep the old behaviour and reduce possibility of regressions.